src/ompl/geometric/planners/rrt/src/RRTstar.cpp

```diff
@@ -573,73 +661,210 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
+    // The number pruned
+    int numPruned = 0;
+
+    if (opt_->isFinite(prunedCost_))
     {
-        Motion *candidate = pruneScratchSpace_.candidates[j++];
-        if (opt_->isCostBetterThan(pruneTreeCost, costToGo(candidate)))
-            pruneScratchSpace_.toBePruned.push_back(candidate);
-        else
+        fracBetter = std::abs((pruneTreeCost.value() - prunedCost_.value())/prunedCost_.value());
+    }
+    else
+    {
+        fracBetter = 1.0;
+    }
+
+    if (fracBetter > pruneThreshold_)
+    {
+        // We are only pruning motions if they, AND all descendents, meet the pruning criterion
+        // The easiest way to do this is to find all the leaves of the tree that meet the pruning criterion, and ascend up the tree from each leaf until a motion is found that should not be pruned
+
+        // Variable
+        // The list of motions in the NN structure, initially full of nulls
+        std::vector<Motion*> treeVertices(nn_->size(), NULL);
+        // The list of leaves to prune
+        std::list<Motion*> leavesToPrune;
+
+        // Get the list from the NN structure
+        nn_->list(treeVertices);
+
+        // Iterate over the list of vertices
+        for (std::vector<Motion*>::iterator vIter = treeVertices.begin(); vIter != treeVertices.end(); ++vIter)
+        {
+            // Check if it passes the pruning condition
+            if (leafPruningCondition(*vIter, pruneTreeCost) == true)
+            {
+                // Add to the list of leaves to prune
+                leavesToPrune.push_back(*vIter);
+            }
+            // No else, does not meet the pruning criteria
+        }
+
+        // Iterate over the list of leaves to prune
+        std::list<Motion*>::iterator lpIter = leavesToPrune.begin();
+
+        while (lpIter != leavesToPrune.end())
         {
-            pruneScratchSpace_.newTree.push_back(candidate);
-            pruneScratchSpace_.candidates.insert(pruneScratchSpace_.candidates.end(),
-                candidate->children.begin(), candidate->children.end());
+            // Remove the leaf from the nn_ structure
+            nn_->remove(*lpIter);
+
+            // Remove the leaf from its parent
+            removeFromParent(*lpIter);
+
+            // Check if its parent should be added to the list
+            if (leafPruningCondition((*lpIter)->parent, pruneTreeCost) == true)
+            {
+                // Add to the list of leaves to prune
+                leavesToPrune.push_back((*lpIter)->parent);
+            }
+            // No else, does not meet the pruning criteria
+
+            // Erase the actual motion, by first freeing the state
+            si_->freeState((*lpIter)->state);
+
+            // then deleting the pointer
+            delete *lpIter;
+
+            // And finally removing it from the list, erase returns the next iterator
+            lpIter = leavesToPrune.erase(lpIter);
+
+            // Update our counter
+            ++numPruned;
+        }
```

Comment:
This iterates through the list of leaves to trim, working from the front, we

1. Remove the leaf from the NN struct
2. Disconnecting it from its parent
3. Check if the parent should now be added to the list, (i.e., is it now a leaf? is it's heuristic estimate too high?) and adding if necessary.
4. Freeing the state and pointer
5. Removing from the list and getting the next iterator.